lib/std/DoublyLinkedList.zig

```diff
@@ -0,0 +1,284 @@
+pub const Node = struct {
+    prev: ?*Node = null,
+    next: ?*Node = null,
+};
+
+pub fn insertAfter(list: *DoublyLinkedList, existing_node: *Node, new_node: *Node) void {
+    new_node.prev = existing_node;
+    if (existing_node.next) |next_node| {
+        // Intermediate node.
+        new_node.next = next_node;
+        next_node.prev = new_node;
+    } else {
+        // Last element of the list.
+        new_node.next = null;
+        list.last = new_node;
+    }
+    existing_node.next = new_node;
+}
+
+pub fn insertBefore(list: *DoublyLinkedList, existing_node: *Node, new_node: *Node) void {
+    new_node.next = existing_node;
+    if (existing_node.prev) |prev_node| {
+        // Intermediate node.
+        new_node.prev = prev_node;
+        prev_node.next = new_node;
+    } else {
+        // First element of the list.
+        new_node.prev = null;
+        list.first = new_node;
+    }
+    existing_node.prev = new_node;
+}
+
+/// Concatenate list2 onto the end of list1, removing all entries from the former.
+///
+/// Arguments:
+///     list1: the list to concatenate onto
+///     list2: the list to be concatenated
+pub fn concatByMoving(list1: *DoublyLinkedList, list2: *DoublyLinkedList) void {
+    const l2_first = list2.first orelse return;
+    if (list1.last) |l1_last| {
+        l1_last.next = list2.first;
+        l2_first.prev = list1.last;
+    } else {
+        // list1 was empty
+        list1.first = list2.first;
+    }
+    list1.last = list2.last;
+    list2.first = null;
+    list2.last = null;
+}
+
+/// Insert a new node at the end of the list.
+///
+/// Arguments:
+///     new_node: Pointer to the new node to insert.
+pub fn append(list: *DoublyLinkedList, new_node: *Node) void {
+    if (list.last) |last| {
+        // Insert after last.
+        list.insertAfter(last, new_node);
+    } else {
+        // Empty list.
+        list.prepend(new_node);
+    }
+}
+
+/// Insert a new node at the beginning of the list.
+///
+/// Arguments:
+///     new_node: Pointer to the new node to insert.
+pub fn prepend(list: *DoublyLinkedList, new_node: *Node) void {
+    if (list.first) |first| {
+        // Insert before first.
+        list.insertBefore(first, new_node);
+    } else {
+        // Empty list.
+        list.first = new_node;
+        list.last = new_node;
+        new_node.prev = null;
+        new_node.next = null;
+    }
+}
+
+/// Remove a node from the list.
+///
+/// Arguments:
+///     node: Pointer to the node to be removed.
+pub fn remove(list: *DoublyLinkedList, node: *Node) void {
+    if (node.prev) |prev_node| {
+        // Intermediate node.
+        prev_node.next = node.next;
+    } else {
+        // First element of the list.
+        list.first = node.next;
+    }
+
+    if (node.next) |next_node| {
+        // Intermediate node.
+        next_node.prev = node.prev;
+    } else {
+        // Last element of the list.
+        list.last = node.prev;
```

Comment:
What if the node is a member of a different list? I dont see how that is asserted here.

Reply:
It isn't asserted here, and the old code didn't either, so I think it's off topic for this PR.
https://github.com/ziglang/zig/pull/23459/files#diff-b9b128f3e3b9a07cbfb6d5e388e1fee160ba8fcee363f80f74144c8c8dfa2512L300